002-es6-modules.md

```diff
@@ -0,0 +1,274 @@
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
+    - A means to create a shared binding (variable) from an export to an import. Required for the "live" binding of imports.
+
+* [ModuleNamespaceCreate](https://tc39.github.io/ecma262/#sec-modulenamespacecreate)
+    - Provides a means of creating a list of exports manually, used so that CommonJS `module.exports` can create ModuleRecords that are prepopulated.
+
+---
+
+### [WhatWG Loader](https://github.com/whatwg/loader)
+
+---
+
+Discusses the design of module metadata in a [Registry](https://whatwg.github.io/loader/#registry). All actions regarding the Registry are synchronous.
+
+**NOTE** It is not Node's intent to implement the asynchronous pipeline in the Loader specification. There is discussion about including a synchronous pipeline in the specification as an addendum.
+
+
+### [Summary Video on Youtube](youtube.com/watch?v=NdOKO-6Ty7k)
+
+## Additional Structures Required
+
+### DynamicModuleRecord
+
+A Module Record that presents a view of an Object for its `[[Namespace]]` rather than coming from an environment record.
+
+The list of exports is frozen upon construction. No new properties may be added. No properties may be removed.
+
+## Algorithm
+
+When `require()`ing a file.
+
+1. Determine if file is ES6 or CommonJS (CJS).
+2. If CJS
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES6
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing `require` recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+
+## Semantics
+
+### Determining if source is an ES6 Module
+
+If a module can be parsed outside of the ES6 Module goal, it will be treated as a [Script](https://tc39.github.io/ecma262/#sec-parse-script). Otherwise it will be parsed as [ES6](https://tc39.github.io/ecma262/#sec-parsemodule).
+
+In pseudocode, it looks somewhat like:
+
+```
+try {
+  v8::Script::Compile
+}
+catch (e) {
+  v8::Script::CompileModule
+}
+```
+
+V8 may or may not choose to use parser fallback to combine this into one step.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`.
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+require('./es6');
+// {default:'my-default'}
+```
+
+#### read only
+
+The objects create by an ES6 module are [ModuleNamespace Objects](https://tc39.github.io/ecma262/#sec-module-namespace-exotic-objects).
+
+These have `[[Set]]` be a no-op and are read only views of the exports of an ES6 module.
+
+### ES6 consuming CommonJS
```

Comment:
> One must be able to require('es6') and import foo from 'cjs' without knowing the module format used. In browsers

Really? Why? When someone uses a library it's pretty clear what format that library uses, why do you think user of library may not be aware of its format?

> but that's not the way node (or the existing CJS ecosystem) can, or will, reasonably work.

Why? Node.js may still expect CJS modules at runtime, and accept ES modules via option as e.g. `node --esmodule es-module-program,js`. Then with some major version of Node it may change, and default would be to favor ES module (while CJS being an option). I don't see how it can be an issue.

> The need for browser/node/cjs/es6 interop is clear

ES6 module that requires CJS, indeed will not be able to run natively without recompilation of both CJS module and ES module. I  can see that this can be perceived an issue, however what you propose still won't run without recompilation of CJS module, so as long some module is written as CJS compilation step is mandatory.

> A file extension is, so far, the best option we have that meets the criteria

It's really a call to change '.js' to something else (in long term), it will just increase disruption in ecosystem, and seems as totally not needed (as pointed above).